packages/discovery-provider/plugins/pedalboard/apps/backfill-audio-analyses/src/backfill_discovery.ts

```diff
@@ -0,0 +1,223 @@
+import axios from 'axios'
+import { Semaphore } from 'await-semaphore'
+import { Knex } from 'knex'
+import { App } from '@pedalboard/basekit'
+import { config } from '.'
+import { SharedData } from './index'
+import {
+  storeDbOffset,
+  readDbOffset,
+  getCachedHealthyContentNodes
+} from './redis'
+
+const Web3 = require('web3')
+const web3 = new Web3()
+
+// concurrency control
+const MAX_CONCURRENT_REQUESTS = 10
+const semaphore = new Semaphore(MAX_CONCURRENT_REQUESTS)
+
+const REQUEST_TIMEOUT = 5000 // 5s
+
+const DB_OFFSET_KEY = 'discovery:backfill_audio_analyses:offset'
+
+interface Track {
+  track_id: number
+  track_cid: string
+  audio_upload_id: string | null
+  musical_key: string | null
+  bpm: number | null
+  audio_analysis_error_count: number | null
+}
+
+function formatErrorLog(
+  message: string,
+  track: Track,
+  node: string,
+  attemptNo: number
+): string {
+  let errLog = `Error retrieving audio analysis on ${node}: ${message}. Attempt #${attemptNo} for track ID ${track.track_id}, track CID ${track.track_cid}`
+  if (track.audio_upload_id) {
+    errLog += `, upload ID ${track.audio_upload_id}`
+  }
+  if (attemptNo < 5) {
+    errLog += '. Trying another content node...'
+  } else {
+    errLog += '. Skipping track...'
+  }
+  return errLog
+}
+
+async function getAudioAnalysis(contentNodes: string[], track: Track) {
+  const audioUploadId = track.audio_upload_id || ''
+  const trackCid = track.track_cid
+  if (!trackCid) return
+  const isLegacyTrack = !audioUploadId
+
+  const release = await semaphore.acquire() // acquire a semaphore permit
+
+  let result = null
+  // allow up to 5 attempts to get audio analysis for this track
+  for (let i = 0; i < 5; i++) {
+    // choose a random content node
+    const contentNode =
+      contentNodes[Math.floor(Math.random() * contentNodes.length)]
+    try {
+      let analysisUrl = `${contentNode}/uploads/${audioUploadId}`
+      if (isLegacyTrack) {
+        analysisUrl = `${contentNode}/tracks/legacy/${trackCid}/analysis`
+      }
+
+      const response = await axios.get(analysisUrl, {
+        timeout: REQUEST_TIMEOUT
+      })
+      if (response.status == 200) {
+        console.log(
+          `Successfully retrieved audio analysis for track ID ${
+            track.track_id
+          }, track CID ${trackCid}${
+            audioUploadId ? `, upload ID: ${audioUploadId}` : ''
+          } via ${contentNode}`
+        )
+        const resultsKey = isLegacyTrack ? 'results' : 'audio_analysis_results'
+        const errorCountKey = isLegacyTrack
+          ? 'error_count'
+          : 'audio_analysis_error_count'
+        const results = response.data[resultsKey]
+        const errorCount = response.data[errorCountKey]
+        if (!results) {
+          break
+        }
+
+        let musicalKey = null
+        let bpm = null
+        if (results?.key) {
+          musicalKey = results?.key
+        } else if (results?.Key) {
```

Comment:
interesting...

Reply:
haha yeah i forgot to specify the json tag in the go struct so some early records have the keys as `Key` and `BPM`. is only relevant for the backfill; not for new uploads